lib/src/protobuf/coded_buffer_writer.dart

```diff
@@ -4,221 +4,494 @@
+  int _copyInto(Uint8List buffer, int pos, TypedData value) {
+    if (value is Uint8List) {
+      int len = value.length;
+      for (int j = 0; j < len; j++) {
+        buffer[pos++] = value[j];
+      }
+      return pos;
+    } else {
+      int len = value.lengthInBytes;
+      Uint8List u8 = new Uint8List.view(
+          value.buffer, value.offsetInBytes, value.lengthInBytes);
+      for (int j = 0; j < len; j++) {
+        buffer[pos++] = u8[j];
+      }
+      return pos;
     }
-    return true;
   }
+
+  /// This function maps a power-of-2 value (2^0 .. 2^31) to a unique value
+  /// in the 0..31 range.
+  ///
+  /// For more details see "Using de Bruijn Sequences to Index a 1 in
+  /// a Computer Word"[1]
+  ///
+  /// Note: this is guaranteed to work after compilation to JavaScript
+  /// where multiplication becomes a floating point multiplication.
+  ///
+  /// [1] http://supertech.csail.mit.edu/papers/debruijn.pdf
+  static int _valueTypeIndex(int powerOf2) =>
+      ((0x077CB531 * powerOf2) >> 27) & 31;
+
+  /// Precomputed indices for all FbFieldType._XYZ_BIT values:
+  ///
+  ///    _XYZ_BIT_INDEX = _valueTypeIndex(FbFieldType._XYZ_BIT)
+  ///
+  static const _BOOL_BIT_INDEX = 14;
+  static const _BYTES_BIT_INDEX = 29;
+  static const _STRING_BIT_INDEX = 27;
+  static const _DOUBLE_BIT_INDEX = 23;
+  static const _FLOAT_BIT_INDEX = 15;
+  static const _ENUM_BIT_INDEX = 31;
+  static const _GROUP_BIT_INDEX = 30;
+  static const _INT32_BIT_INDEX = 28;
+  static const _INT64_BIT_INDEX = 25;
+  static const _SINT32_BIT_INDEX = 18;
+  static const _SINT64_BIT_INDEX = 5;
+  static const _UINT32_BIT_INDEX = 11;
+  static const _UINT64_BIT_INDEX = 22;
+  static const _FIXED32_BIT_INDEX = 13;
+  static const _FIXED64_BIT_INDEX = 26;
+  static const _SFIXED32_BIT_INDEX = 21;
+  static const _SFIXED64_BIT_INDEX = 10;
+  static const _MESSAGE_BIT_INDEX = 20;
+
+  /// Mapping from value types to wire-types indexed by _valueTypeIndex(...).
+  static final Uint8List _wireTypes = new Uint8List(32)
+    ..[_BOOL_BIT_INDEX] = WIRETYPE_VARINT
+    ..[_BYTES_BIT_INDEX] = WIRETYPE_LENGTH_DELIMITED
+    ..[_STRING_BIT_INDEX] = WIRETYPE_LENGTH_DELIMITED
+    ..[_DOUBLE_BIT_INDEX] = WIRETYPE_FIXED64
+    ..[_FLOAT_BIT_INDEX] = WIRETYPE_FIXED32
+    ..[_ENUM_BIT_INDEX] = WIRETYPE_VARINT
+    ..[_GROUP_BIT_INDEX] = WIRETYPE_START_GROUP
+    ..[_INT32_BIT_INDEX] = WIRETYPE_VARINT
+    ..[_INT64_BIT_INDEX] = WIRETYPE_VARINT
+    ..[_SINT32_BIT_INDEX] = WIRETYPE_VARINT
+    ..[_SINT64_BIT_INDEX] = WIRETYPE_VARINT
+    ..[_UINT32_BIT_INDEX] = WIRETYPE_VARINT
+    ..[_UINT64_BIT_INDEX] = WIRETYPE_VARINT
+    ..[_FIXED32_BIT_INDEX] = WIRETYPE_FIXED32
+    ..[_FIXED64_BIT_INDEX] = WIRETYPE_FIXED64
+    ..[_SFIXED32_BIT_INDEX] = WIRETYPE_FIXED32
+    ..[_SFIXED64_BIT_INDEX] = WIRETYPE_FIXED64
+    ..[_MESSAGE_BIT_INDEX] = WIRETYPE_LENGTH_DELIMITED;
 }
+
+int _encodeZigZag32(int value) => (value << 1) ^ (value >> 31);
```

Comment:
This might be broken on dart2js since bitops and shifts are basically unsigned-to-unsigned operations.
An alternative:
(value<<1) ^ (0 - ((value>>32)&1))

Is there a zigzag end-to-end test?

Reply:
Seems to work correctly for _in range integers_: because `>>` still does sign extension for negative numbers (even though it then converts result to unsigned value). 

There are tests for this by they leave behavior of things like: "what should happen if you write `0x8000000` as signed int32" - undefined.